docker/Dockerfile.build

```diff
@@ -1,56 +1,88 @@
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
+    apt-get install -y --no-install-recommends \
+    llvm libclang-dev clang ca-certificates && \
+    rm -rf /var/lib/apt/lists/* /tmp/*
 
-RUN mkdir /zebra
-WORKDIR /zebra
+# Optimize builds. In particular, regenerate-stateful-test-disks.yml was reaching the
+# GitHub Actions time limit (6 hours), so we needed to make it faster.
+ENV RUSTFLAGS -O
+ENV CARGO_HOME /app/.cargo/
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
 
-ARG SHORT_SHA
-ENV SHORT_SHA $SHORT_SHA
+# Build zebra
+ARG RUST_BACKTRACE=1
+ENV RUST_BACKTRACE ${RUST_BACKTRACE}
 
-ENV RUST_BACKTRACE full
-ENV CARGO_HOME /zebra/.cargo/
+COPY . .
+RUN cargo build --locked --release --features enable-sentry --bin zebrad
 
-RUN rustc -V; cargo -V; rustup -V
+FROM builder AS tester
+# Skip IPv6 tests by default, as some CI environment don't have IPv6 available
+ARG ZEBRA_SKIP_IPV6_TESTS=1
+ENV ZEBRA_SKIP_IPV6_TESTS ${ZEBRA_SKIP_IPV6_TESTS}
 
-COPY . .
+RUN cargo test --locked --release --features enable-sentry --workspace --no-run
+# Pre-download Zcash Sprout and Sapling parameters
+RUN cargo run --locked --release --features enable-sentry --bin zebrad download
 
-RUN cd zebrad/; cargo build --release --features enable-sentry
+CMD ["cargo" "test" "--locked" "--release" "--features" "enable-sentry" "--workspace"]
 
 # Runner image
-FROM debian:buster-slim AS zebrad-release
+FROM debian:bullseye-slim AS runtime
 
 RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	ca-certificates
+    apt-get install -y --no-install-recommends \
+    ca-certificates
 
-COPY --from=builder /zebra/target/release/zebrad /
+COPY --from=builder /app/target/release/zebrad /usr/local/bin
 
 ARG CHECKPOINT_SYNC=true
 ARG NETWORK=Mainnet
 
-RUN printf "[consensus]\n" >> /zebrad.toml
-RUN printf "checkpoint_sync = ${CHECKPOINT_SYNC}\n" >> /zebrad.toml
-RUN printf "[metrics]\n" >> /zebrad.toml
-RUN printf "endpoint_addr = '0.0.0.0:9999'\n" >> /zebrad.toml
-RUN printf "[network]\n" >> /zebrad.toml
-RUN printf "network = '${NETWORK}'\n" >> /zebrad.toml
-RUN printf "[state]\n" >> /zebrad.toml
-RUN printf "cache_dir = '/zebrad-cache'\n" >> /zebrad.toml
-RUN printf "[tracing]\n" >> /zebrad.toml
-RUN printf "endpoint_addr = '0.0.0.0:3000'\n" >> /zebrad.toml
-RUN cat /zebrad.toml
+RUN set -ex; \
+  { \
+    echo "[consensus]"; \
+    echo "checkpoint_sync = ${CHECKPOINT_SYNC}"; \
+    echo "[metrics]"; \
+    echo "endpoint_addr = '0.0.0.0:9999'"; \
+    echo "[network]"; \
+    echo "network = '${NETWORK}'"; \
+    echo "[state]"; \
+    echo "cache_dir = '/zebrad-cache'"; \
+    echo "[tracing]"; \
+    echo "endpoint_addr = '0.0.0.0:3000'"; \
+  } > "zebrad.toml"
+
+RUN cat zebrad.toml
 
 # Pre-download Zcash Sprout and Sapling parameters
-RUN /zebrad download
+RUN /usr/local/bin/zebrad download
```

Comment:
Can we copy or re-use the parameters from the tests, rather than downloading them?

We're trying to reduce the number of CI failures due to parameter download rate-limits: #3447

Reply:
How frequent does this parameters change state? Just to be sure if there's a better solution to this.

Reply:
The parameters are fixed and will never change.

However, we need to re-test the download process every time the download code changes: #3448
(But we don't need to do that in this specific workflow, we could add a CI job for it.)

Reply:
Downloading just once, and copying the results to the following builds.